cpu/esp32/periph/spi.c

```diff
@@ -263,11 +263,66 @@ void spi_deinit_pins(spi_t bus)
     mutex_lock(&_spi[bus].lock);
 }
 
+spi_clk_t IRAM_ATTR spi_get_clk(spi_t bus, uint32_t freq)
+{
+    (void)bus;
+    /*
+     * set SPI clock
+     * see ESP32 Technical Reference, Section 7.8 SPI_CLOCK_REG
+     * https://www.espressif.com/sites/default/files/documentation/esp32_technical_reference_manual_en.pdf
+     */
+
+    uint32_t apb_clk = rtc_clk_apb_freq_get();
+    uint32_t clk_reg;
+
+    /* Section 7.4 GP­SPI Clock Control:
+     * f_spi = f_apb / ((spi_clkcnt_n + 1) * (spi_clkdiv_pre + 1))
+     * spi_clkdiv_pre 0..8191 (13 bit)
+     * spi_clkcnt_N 1..63 (6 bit)
+     */
+    if (freq > apb_clk / 5) {
+        LOG_TAG_ERROR("spi", "APB clock rate (%"PRIu32" Hz) has to be at "
+                      "least 5 times SPI clock rate (%"PRIu32" Hz)\n",
+                      apb_clk, freq);
```

Comment:
@gschorcht I moved this code as is from spi_acquire but I wonder why APB clock rate has to be at least 5 times SPI clock rate. From the reference manual section 7.4: _"The maximum output clock frequency of ESP32 GP-SPI master is f apb /2"_ and even _"When the SPI_CLK_EQU_SYSCLK bit in register SPI_CLOCK_REG is set to 1, and the other bits are set to 0, SPI output clock frequency is f apb."_

Reply:
This is what I had to learn when I was implementing the support for core clock frequencies lower than 80 MHz down to 2 MHz. For example with
```
CFLAGS='-DCONFIG_ESP32_DEFAULT_CPU_FREQ_MHZ=40' BOARD=esp32-wroom-32 make -j8 -C tests/periph/spi flash
```
and wired MOSI-MISO I get for 10 MHz:
```
> init 0 0 4 0 5
> send hallo
Sent bytes
   0    1    2    3    4 
  0x68 0x65 0x6c 0x6c 0x6f
    h    e    l    l    o 

Received bytes
   0    1    2    3    4 
  0x34 0x32 0xb6 0x36 0x37
    4    2   ??    6    7 
```
while it works for 5 MHz.

I have the same problem for 2 MHz core clock frequency. While it works for 400 kHz SPI clock frequency, it doesn't work for 1 MHz SPI clock frequency.

Reply:
Note that 0x3432b63637 = 0x68656c6c6f >> 1. I observed the same issue on esp8266 for the highest frequency (80MHz with spi_clk_equ_sysclk).

It seems to be linked to spi_miso_delay_mode which should be set to 0 for frequencies higher than clkapb/4 according to the manual section 7.4.2.